File: src/main/java/com/fengsheng/card/WuDao.java
package com.fengsheng.card;

import com.fengsheng.*;
import com.fengsheng.phase.FightPhaseIdle;
import com.fengsheng.phase.OnUseCard;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class WuDao extends Card {
    private static final Logger log = Logger.getLogger(WuDao.class);

    public WuDao(int id, Common.color[] colors, Common.direction direction, boolean lockable) {
        super(id, colors, direction, lockable);
    }

    public WuDao(int id, Card card) {
        super(id, card);
    }

    /**
     * 仅用于“作为误导使用”
     */
    WuDao(Card originCard) {
        super(originCard);
    }

    @Override
    public Common.card_type getType() {
        return Common.card_type.Wu_Dao;
    }

    @Override
    public boolean canUse(Game g, Player r, Object... args) {
        if (r == g.getJinBiPlayer()) {
            log.error("你被禁闭了，不能出牌");
            return false;
        }
        if (g.getQiangLingTypes().contains(getType())) {
            log.error("误导被禁止使用了");
            return false;
        }
        Player target = (Player) args[0];
        if (!(g.getFsm() instanceof FightPhaseIdle fsm)) {
            log.error("误导的使用时机不对");
            return false;
        }
        Player left = fsm.inFrontOfWhom.getNextLeftAlivePlayer();
        Player right = fsm.inFrontOfWhom.getNextRightAlivePlayer();
        if (target == fsm.inFrontOfWhom || (target != left && target != right)) {
            log.error("误导只能选择情报当前人左右两边的人作为目标");
            return false;
        }
        return true;
    }

    @Override
    public void execute(final Game g, final Player r, Object... args) {
        Player target = (Player) args[0];
        log.info(r + "对" + target + "使用了" + this);
        final var fsm = (FightPhaseIdle) g.getFsm();
        r.deleteCard(this.id);
        Fsm resolveFunc = () -> {
            fsm.inFrontOfWhom = target;
            fsm.whoseFightTurn = fsm.inFrontOfWhom;
            g.getDeck().discard(this.getOriginCard());
            for (Player player : g.getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    var builder = Fengsheng.use_wu_dao_toc.newBuilder().setCard(this.toPbCard());
                    builder.setPlayerId(p.getAlternativeLocation(r.location()));
                    builder.setTargetPlayerId(p.getAlternativeLocation(target.location()));
                    p.send(builder.build());
                }
            }
            return new ResolveResult(fsm, true);
        };
        g.resolve(new OnUseCard(fsm.whoseTurn, r, null, this, Common.card_type.Wu_Dao, r, resolveFunc));
    }

    @Override
    public String toString() {
        return Card.cardColorToString(colors) + "误导";
    }

    public static boolean ai(FightPhaseIdle e, Card card) {
        Player player = e.whoseFightTurn;
        var colors = e.messageCard.getColors();
        if (e.inFrontOfWhom == player && (e.isMessageCardFaceUp || player == e.whoseTurn) && colors.size() == 1 && colors.get(0) != Common.color.Black)
            return false;
        Player target = switch (ThreadLocalRandom.current().nextInt(4)) {
            case 0 -> e.inFrontOfWhom.getNextLeftAlivePlayer();
            case 1 -> e.inFrontOfWhom.getNextRightAlivePlayer();
            default -> null;
        };
        if (target == null) return false;
        final Player finalTarget = target;
        GameExecutor.post(player.getGame(), () -> {
            Card card0 = card.getType() == Common.card_type.Wu_Dao ? card : Card.falseCard(Common.card_type.Wu_Dao, card);
            card0.execute(player.getGame(), player, finalTarget);
        }, 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/card/WuDao.kt
package com.fengsheng.card

import com.fengsheng.Game
import com.fengsheng.HumanPlayer
import com.fengsheng.Player
import com.fengsheng.phase.FightPhaseIdle
import com.fengsheng.phase.OnFinishResolveCard
import com.fengsheng.phase.ResolveCard
import com.fengsheng.protos.Common.card_type.Wu_Dao
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.direction
import com.fengsheng.protos.Common.phase.Fight_Phase
import com.fengsheng.protos.notifyPhaseToc
import com.fengsheng.protos.useWuDaoToc
import com.fengsheng.skill.cannotPlayCard
import org.apache.logging.log4j.kotlin.logger

class WuDao : Card {
    constructor(id: Int, colors: List<color>, direction: direction, lockable: Boolean) :
            super(id, colors, direction, lockable)

    constructor(id: Int, card: Card) : super(id, card)

    /**
     * 仅用于“作为误导使用”
     */
    internal constructor(originCard: Card) : super(originCard)

    override val type = Wu_Dao

    override fun canUse(g: Game, r: Player, vararg args: Any): Boolean {
        if (r.cannotPlayCard(type)) {
            logger.error("你被禁止使用误导")
            (r as? HumanPlayer)?.sendErrorMessage("你被禁止使用误导")
            return false
        }
        val target = args[0] as Player
        val fsm = g.fsm as? FightPhaseIdle
        if (fsm == null) {
            logger.error("误导的使用时机不对")
            (r as? HumanPlayer)?.sendErrorMessage("误导的使用时机不对")
            return false
        }
        val left = fsm.inFrontOfWhom.getNextLeftAlivePlayer()
        val right = fsm.inFrontOfWhom.getNextRightAlivePlayer()
        if (target === fsm.inFrontOfWhom || target !== left && target !== right) {
            logger.error("误导只能选择情报当前人左右两边的人作为目标")
            (r as? HumanPlayer)?.sendErrorMessage("误导只能选择情报当前人左右两边的人作为目标")
            return false
        }
        return true
    }

    override fun execute(g: Game, r: Player, vararg args: Any) {
        val target = args[0] as Player
        logger.info("${r}对${target}使用了$this")
        g.resolve(onUseCard(this, g, r, target))
    }

    override fun toString(): String {
        return "${cardColorToString(colors)}误导"
    }

    companion object {
        /**
         * 执行【误导】的效果。示例：
         * ```
         * g.resolve(WuDao.onUseCard(card, g, r, target))
         * ```
         * @param card 使用的那张【误导】卡牌。可以为 `null` ，因为SP阿芙罗拉技能【应变自如】可以视为使用了【误导】。
         * @return 返回[ResolveCard]，要自行调用[Game.resolve]
         */
        fun onUseCard(card: Card?, g: Game, r: Player, target: Player): ResolveCard {
            val fsm = g.fsm as FightPhaseIdle
            card?.apply { r.deleteCard(id) }
            val resolveFunc = { valid: Boolean ->
                if (valid) {
                    for (player in g.players) {
                        if (player is HumanPlayer) {
                            player.send(useWuDaoToc {
                                card?.let { this.card = it.toPbCard() }
                                playerId = player.getAlternativeLocation(r.location)
                                targetPlayerId = player.getAlternativeLocation(target.location)
                            })
                        }
                    }
                    val newFsm = fsm.copy(inFrontOfWhom = target, whoseFightTurn = target)
                    for (p in g.players) { // 解决客户端动画问题
                        if (p is HumanPlayer) {
                            p.send(notifyPhaseToc {
                                currentPlayerId = p.getAlternativeLocation(newFsm.whoseTurn.location)
                                messagePlayerId = p.getAlternativeLocation(newFsm.inFrontOfWhom.location)
                                waitingPlayerId = p.getAlternativeLocation(newFsm.whoseFightTurn.location)
                                currentPhase = Fight_Phase
                                if (newFsm.isMessageCardFaceUp) messageCard = newFsm.messageCard.toPbCard()
                            })
                        }
                    }
                    OnFinishResolveCard(fsm.whoseTurn, r, target, card?.getOriginCard(), Wu_Dao, newFsm)
                } else {
                    val newFsm = fsm.copy(whoseFightTurn = fsm.inFrontOfWhom)
                    OnFinishResolveCard(fsm.whoseTurn, r, target, card?.getOriginCard(), Wu_Dao, newFsm)
                }
            }
            return ResolveCard(
                fsm.whoseTurn, r, target, card?.getOriginCard(), Wu_Dao, resolveFunc, fsm,
                valid = target !== fsm.inFrontOfWhom
            )
        }
    }
}